happi/audit.py

```diff
@@ -0,0 +1,218 @@
+        """
+        dict((key, self.create_arg(item, val))
+             for key, val in item.kwargs.items())
+
+    def create_arg(self, item, arg):
+        """
+        Function borrowed from loader to create
+        correctly typed arguments from happi information
+        """
+        if not isinstance(arg, str):
+            return arg
+        try:
+            return fill_template(arg, item, enforce_type=True)
+        except Exception as e:
+            logger.warning('Probably provided invalid argument: %s for %s, %s',
+                           arg, item.name, e)
+
+    def validate_device_class(self, item):
+        """
+        Validates device_class field
+        """
+        device_class = item.get('device_class')
+        device = item.get('name')
+        if not device_class:
+            logger.warning('Detected a None vlaue for %s. '
+                           'The device_class cannot be None', device)
+        else:
+            try:
+                mod, cls = device_class.rsplit('.', 1)
+            except (Exception, ValueError) as e:
+                logger.warning('Wrong device name format: %s for %s, %s',
+                               device_class, device, e)
+            else:
+                try:
+                    import_class(device_class)
+                except ImportError as ex:
+                    logger.warning(ex)
+
+    def validate_extra_attributes(self, item):
+        """
+        Validate items that have extra attributes
+        """
+        attr_list = []
+        extr_list = ['creation', 'last_edit', '_id', 'type']
+
+        for (key, value), s in zip(item.items(), item.entry_info):
+            attr_list.append(s.key)
+
+        key_list = [key for key, value in item.items()]
+        diff = set(key_list) - set(attr_list) - set(extr_list)
+        if diff:
+            logger.warning('Device %s has extra attributes %s',
+                           item.name, diff)
+
+    def print_report_message(self, message):
+        logger.info('')
+        logger.info('--------- %s ---------', message)
+        logger.info('')
+
+    def parse_database(self, database_path):
+        """
+        Validates if an entry is a valid entry in the database
+
+        Parameters
+        ----------
+        database_path: str
+            Path to the database to be validated
+
+        """
+        client = Client(path=database_path)
+        items = client.all_items
+
+        self.print_report_message('VALIDATING ENTRIES')
+        client.validate()
+
+        self.print_report_message('VALIDATING ARGS, KWARGS, EXTRA ATTRIBUTES')
+        for item in items:
+            self.validate_args(item)
+            self.validate_kwargs(item)
+            self.validate_extra_attributes(item)
+            pass
+
+        # TODO: what to do here???....
+        # for entr, item in zip(items, client.backend.all_devices):
+        #     #self.validate_enforce(entr, item)
+        #     pass
```

Comment:
I think https://github.com/pcdshub/happi/blob/c77ed7d28c74d946800afad5a60ced5f9e317500/happi/item.py#L82 is the key function to use here, to check that the values saved in the item match their respective EntryInfo

Reply:
Yes I see what you mean, I am testing against `enforce_value` already - i like it, and it works great as long as we only add to the database using `happi` :) 
If for example I have a number instead of  a string or even a numerical string instead of float - this will still work, but the loading will not work, and I wanted to see if I can catch that - but if I already initialize an EntryInfo, it is too late at that point to check for that I think, it has to be happening before even initializing it.....

Reply:
I'm not sure I fully understand. Are you saying that if someone manually edits the json, an enforce_value can pass, despite being in a form that fails in the loader later?

Reply:
Maybe I got it wrong here, i need to revisit this one, but for example if i try to add this item:  **Please note that i have a string for the `z` value**

```
device_info = {'name': 'alias3',
            'z': '400',
            '_id': 'alias3',
            'prefix': 'BASE:PV',
            'beamline': 'LCLS',
            'type': 'OphydItem',
            'device_class': 'types.SimpleNamespace',
            'args': list(),
            'kwargs': {'hi': 'oh hello'},
            'location_group': 'LOC',
            'functional_group': 'FUNC',
            }
device = OphydItem(**device_info)
client.add_device(device)

```
This will add fine, will also validate fine, and even load fine - but we might have problems when we look for this z value and try to search it... 

One other example I can think of is this one: **Note the beamline == True** - this will convert fine to a string so it won't give us issues saving it, or even loading it - but technically it is not a valid entry??

```
device_info = {'name': 'alias4',
            'z': '400',
            '_id': 'alias4',
            'prefix': 'BASE:PV',
            'beamline': True,
            'type': 'OphydItem',
            'device_class': 'types.SimpleNamespace',
            'args': list(),
            'kwargs': {'hi': 'oh hello'},
            'location_group': 'LOC',
            'functional_group': 'FUNC',
            }
```

Maybe I am being a bit too much here though hahah - please let me know :)

Reply:
It seems like `happi` is being very permissive for a lot of these fields: as long as you can cast it to the required type, it is ok as far as `happi` is concerned. I haven't thought for long enough if this is something that should be supported.

If this is causing issues with `search` then we clearly have to intervene a bit... I'm not sure the best resolution. It sounds just as much like a `search` bug as a validation error though.

Reply:
I agree, I think as long as we can cast it to the correct type we should be fine - which we already do, and also if we have problems searching, this should be part of the searching logic to handle. Thank you.